alacritty_terminal/src/term/mod.rs

```diff
@@ -1368,9 +1319,71 @@ impl Term {
     pub fn clipboard(&mut self) -> &mut Clipboard {
         &mut self.clipboard
     }
+
+    pub fn urls(&self) -> Vec<Url> {
+        let display_offset = self.grid.display_offset();
+        let num_cols = self.grid.num_cols().0;
+        let last_col = Column(num_cols - 1);
+        let last_line = self.grid.num_lines() - 1;
+
+        let grid_end_point = Point::new(0, last_col);
+        let mut iter = self.grid.iter_from(grid_end_point);
+
+        let mut parser = Parser::new();
+        let mut extra_url_len = 0;
+        let mut urls = Vec::new();
+
+        while let Some(cell) = iter.prev() {
+            // Skip double-width cell but extend URL length
+            if cell.flags.contains(cell::Flags::WIDE_CHAR_SPACER) {
+                extra_url_len += 1;
+                continue;
+            }
+
+            // Interrupt URLs on line break
+            if iter.cur.col == last_col && !cell.flags.contains(cell::Flags::WRAPLINE) {
+                extra_url_len = 0;
+                parser.reset();
+            }
+
+            match parser.advance(cell.c) {
+                ParserState::Url(length) => {
+                    urls.push(Url::new(iter.cur, length + extra_url_len, num_cols))
+                },
+                ParserState::NoUrl => {
+                    extra_url_len = 0;
+
+                    // Stop searching for URLs once the viewport has been left without active URL
+                    if iter.cur.line > last_line.0 + display_offset {
+                        break;
+                    }
+                },
+                _ => (),
+            }
+        }
+
+        urls
+    }
+
+    pub fn url_to_string(&self, url: &Url) -> String {
+        let mut url_text = String::new();
+
+        let mut iter = self.grid.iter_from(url.start);
+        loop {
```

Comment:
I think this function can be much simpler and expressed entirely in iterators. Oh, and you're skipping first char here you can add `println!("URL: {}", url_text)`, so opening `https://github.com/chrisduerr/rfind_url` results in `URL: ttps://github.com/chrisduerr/rfind_url`

Reply:
> opening https://github.com/chrisduerr/rfind_url results in URL: ttps://github.com/chrisduerr/rfind_url

This explains so much. I've actually extended the URL in another place because I thought there was an off by one issue, but it seems like this is the problem.

Doesn't help that Firefox apparently doesn't complain at all when opening `ttps://example.org`.

Reply:
Yeah, it auto adds h to url.

Reply:
Regarding simplification though, I don't think iterators would make this much cleaner. Especially since we need a reference to the iterator to make sure we can return early.

Reply:
> Regarding simplification though, I don't think iterators would make this much cleaner. Especially since we need a reference to the iterator to make sure we can return early.

After looking at this function once again, making it correct probably will be enough here.